walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,102 @@
 
 package com.google.sps;
 
-import java.util.Collection;
+import java.util.*;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        // Get necessary info from the MeetingRequest object.
+        Collection<String> requiredAttendees = request.getAttendees();
+        Collection<String> optionalAttendees = request.getOptionalAttendees();
+        long duration = request.getDuration();
+
+        // Get all the timeRanges where the attendees have scheduled meetings.
+        ArrayList<TimeRange> requiredTimes = getEventTimes(events, requiredAttendees);
+        ArrayList<TimeRange> optionalTimes = getEventTimes(events, optionalAttendees);
+
+        ArrayList<TimeRange> allTimes = new ArrayList<TimeRange>(requiredTimes);
+        allTimes.addAll(optionalTimes);
+
+        // Sort the meeting time intervals.
+        requiredTimes = sortTimeRanges(requiredTimes);
+        allTimes = sortTimeRanges(allTimes);
+
+        // Find times where the attendees can meet.
+        Collection<TimeRange> requiredMeetings = new ArrayList<TimeRange>();
+        if(requiredTimes.size() != 0) {
+            requiredMeetings = findMeetingTimes(requiredTimes, duration);
+        }
+        Collection<TimeRange> optionalMeetings = findMeetingTimes(allTimes, duration);
+
+        if(optionalMeetings.size() != 0) {
+            return optionalMeetings;
+        }
+        return requiredMeetings;
+    }
+
+    /**
+     * Return all the times where attendees have scheduled meetings.
+     */
+    private ArrayList<TimeRange> getEventTimes(Collection<Event> events, Collection<String> attendees) {
+        ArrayList<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+        for(Event event : events) {
+            if(isScheduled(event, attendees)) {
+                meetingTimes.add(event.getWhen());
+            }
+        }
+        return meetingTimes;
+    }
+
+    /**
+     * Check if an event has certain attendees scheduled for it.
+     */
+    private boolean isScheduled(Event event, Collection<String> attendees) {
+        Set<String> eventAttendees = event.getAttendees();
+
+        for(String attendee : attendees) {
+            if(eventAttendees.contains(attendee)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Orders TimeRanges by their start-times in ascending order.
+     */
+    private ArrayList<TimeRange> sortTimeRanges(ArrayList<TimeRange> relevantTimes) {
+        Collections.sort(relevantTimes, TimeRange.ORDER_BY_START);
+        return relevantTimes;
+    }
+
+    /**
+     * Find times that the attendees can meet.
```

Comment:
This method seems to assume that the relevantTimes array is sorted. That should be mentioned in the javadoc

Reply:
Also "relevantTimes" is a bit uninformative as a variable name.
"relevant" is implied by the fact that it's being passed to the method

Reply:
Maybe the function should be called findAvailableTimes(Collection<TimeRange> unavailableTimes, long minimumDuration)?